pkg/operator/ceph/cluster/crash/reconcile.go

```diff
@@ -0,0 +1,161 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/operator/ceph/disruption/controllerconfig"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	logger = capnslog.NewPackageLogger("github.com/rook/rook", controllerName)
+	// Implement reconcile.Reconciler so the controller can reconcile objects
+	_ reconcile.Reconciler = &ReconcileNode{}
+)
+
+// ReconcileNode reconciles ReplicaSets
+type ReconcileNode struct {
+	// client can be used to retrieve objects from the APIServer.
+	scheme *runtime.Scheme
+	client client.Client
+}
+
+// Reconcile reconciles a node and ensures that it has a crashcollector deployment
+// attached to it.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileNode) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// workaround because the rook logging mechanism is not compatible with the controller-runtime logging interface
+	result, err := r.reconcile(request)
+	if err != nil {
+		logger.Error(err)
+	}
+	return result, err
+}
+
+func (r *ReconcileNode) reconcile(request reconcile.Request) (reconcile.Result, error) {
+
+	logger.Debugf("reconciling node: %q", request.Name)
+
+	// get the node object
+	node := &corev1.Node{ObjectMeta: metav1.ObjectMeta{Name: request.Name}}
+	err := r.client.Get(context.TODO(), request.NamespacedName, node)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get node %q", request.NamespacedName)
+	}
+
+	// Get the list of all the Ceph pods
+	cephPods, err := r.cephPodList()
+	if err != nil {
+		if len(cephPods) == 0 {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to list all ceph pods. %+v", err)
+	}
+
+	namespaceToPodList := make(map[string][]corev1.Pod)
+	for _, cephPod := range cephPods {
+		podNamespace := cephPod.GetNamespace()
+		podList, ok := namespaceToPodList[podNamespace]
+		if !ok {
+			// initialize list
+			namespaceToPodList[podNamespace] = []corev1.Pod{cephPod}
+		} else {
+			// append cephPod to namespace's pod list
+			namespaceToPodList[podNamespace] = append(podList, cephPod)
+		}
+	}
+
+	for namespace, cephPods := range namespaceToPodList {
+		// get dataDirHostPath from the CephCluster
+		cephClusters := &cephv1.CephClusterList{}
+		err := r.client.List(context.TODO(), cephClusters, client.InNamespace(namespace))
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("could not get cephcluster in namespaces %q. %+v", namespace, err)
+		}
+		if len(cephClusters.Items) < 1 {
+			logger.Debugf("no CephCluster found in the namespace %q", namespace)
+			return reconcile.Result{}, nil
+		}
+
+		cephCluster := cephClusters.Items[0]
+		if len(cephClusters.Items) > 1 {
```

Comment:
Should we loop through until we find the first one with the `dataDirHostPath` set? If the first one is a spec for an external cluster it won't have the host path set, right?

Reply:
We can't survive more than one CephCluster in a namespace anyway right? I don't think that will help.

Reply:
makes sense, i missed that we only queried in the namespace

Reply:
I don't remember exactly why but external cluster needs `dataDirHostPath` too.